library/core/src/future/async_drop.rs

```diff
@@ -0,0 +1,269 @@
+/// on a value and its fields recursively.
+#[unstable(feature = "async_drop", issue = "none")]
+pub async fn async_drop<T>(value: T) {
+    let mut value = MaybeUninit::new(value);
+    // SAFETY: value pointer stays valid when needed
+    unsafe { async_drop_in_place(value.as_mut_ptr()) }.await;
+}
+
+#[lang = "async_drop_in_place"]
+#[allow(unconditional_recursion)]
+// FIXME: Consider if `#[rustc_diagnostic_item = "ptr_drop_in_place"]` is needed?
+unsafe fn async_drop_in_place_raw<T: ?Sized>(
+    to_drop: *mut T,
+) -> <T as AsyncDestruct>::AsyncDestructor {
+    // Code here does not matter - this is replaced by the
+    // real async drop glue constructor by the compiler.
+
+    // SAFETY: see comment above
+    unsafe { async_drop_in_place_raw(to_drop) }
+}
+
+/// Creates the asynchronous destructor of the pointed-to value.
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `to_drop` must be [valid](crate::ptr#safety) for both reads and writes.
+///
+/// * `to_drop` must be properly aligned, even if `T` has size 0.
+///
+/// * `to_drop` must be nonnull, even if `T` has size 0.
+///
+/// * The value `to_drop` points to must be valid for async dropping,
+///   which may mean it must uphold additional invariants. These
+///   invariants depend on the type of the value being dropped. For
+///   instance, when dropping a Box, the box's pointer to the heap must
+///   be valid.
+///
+/// * While `async_drop_in_place` is executing or the returned async
+///   destructor is alive, the only way to access parts of `to_drop`
+///   is through the `self: Pin<&mut Self>` references supplied to the
+///   `AsyncDrop::async_drop` methods that `async_drop_in_place` or
+///   `AsyncDropInPlace<T>::poll` invokes.
+///
+#[unstable(feature = "async_drop", issue = "none")]
+pub unsafe fn async_drop_in_place<T: ?Sized>(to_drop: *mut T) -> AsyncDropInPlace<T> {
+    // SAFETY: `async_drop_in_place_raw` has the same safety requirements
+    unsafe { AsyncDropInPlace(async_drop_in_place_raw(to_drop)) }
+}
+
+/// A future returned by the [`async_drop_in_place`].
+#[unstable(feature = "async_drop", issue = "none")]
+// FIXME: Consider if it should be always `!Unpin`
+pub struct AsyncDropInPlace<T: ?Sized>(<T as AsyncDestruct>::AsyncDestructor);
+
+#[unstable(feature = "async_drop", issue = "none")]
+impl<T: ?Sized> fmt::Debug for AsyncDropInPlace<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AsyncDropInPlace").finish_non_exhaustive()
+    }
+}
+
+#[unstable(feature = "async_drop", issue = "none")]
+impl<T: ?Sized> Future for AsyncDropInPlace<T> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // SAFETY: This code simply forwards poll call to the inner future
+        unsafe { self.map_unchecked_mut(|p| &mut p.0).poll(cx) }
+    }
+}
```

Comment:
or just use the projection directly, you can project onto lang items, so you don't need an additional one

Reply:
> or just use the projection directly, you can project onto lang items, so you don't need an additional one

I've tried that but it still didn't let me.

Reply:
What was the issue?

Reply:
> Maybe use a type alias impl trait instead? It won't have the `Debug` impl, but it would avoid having to implement `Future` manually

For alias type definition:

```rust
pub type AsyncDropInPlace<T: ?Sized> = impl Future<Output = ()>;
```

core would require adding to it new `#![feature(type_alias_impl_trait)]` feature gate, and would still fail to compile because compiler would not be able to infer any concrete type due to `async_drop_in_place` definition being recursive (to be eventually replaced by compiler generated shim):

```
unconstrained opaque type
note: `AsyncDropInPlace` must be used in combination with a concrete type within the same module
```